File: ironpythoninaction/chapter14/14.2.1/DynamicObject/DynamicObject/Class1.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.CompilerServices;

namespace DynamicObject
{
    public class DynamicObject
    {
        private Dictionary<string, object> _instance_dict = new Dictionary<string, object>();

        public Dictionary<string, object> container
        {
            get { return _instance_dict; }
        }

        [SpecialName]
        public object GetBoundMember(string name)
        {
            if (!_instance_dict.ContainsKey(name))
            {
                string msg = String.Format("'DynamicObject' has no attribute '{0}'", name);
                throw new System.MissingMemberException(msg);
            } 
            return _instance_dict[name];
        }

        [SpecialName]
        public void SetMemberAfter(string name, object o)
        {
            _instance_dict.Add(name, o);
        }

        [SpecialName]
        public void DeleteMember(string name)
        {
            if (!_instance_dict.ContainsKey(name))
            {
                string msg = String.Format("'DynamicObject' has no attribute '{0}'", name);
                throw new System.MissingMemberException(msg);
            }
            _instance_dict.Remove(name);
        }
    }
}

File: ironpythoninaction/chapter14/14.2.1/DynamicObjectVB/DynamicObjectVB/Class1.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Runtime.CompilerServices

Public Class DynamicObject

    Private _error As String = "'DynamicObject' has no attribute '{0}'"
    Private _instance_dict As New Dictionary(Of String, Object)
    Public ReadOnly Property container() As Dictionary(Of String, Object)
        Get
            Return _instance_dict
        End Get
    End Property

    <SpecialName()> Public Function GetBoundMember(ByVal name As String) As Object
        If Not _instance_dict.ContainsKey(name) Then
            Dim msg As String = String.Format(_error, name)
            Throw New System.MissingMemberException(msg)
        End If
        Return _instance_dict.Item(name)
    End Function

    <SpecialName()> Public Sub SetMemberAfter(ByVal name As String, ByVal value As Object)
        _instance_dict.Add(name, value)
    End Sub

    <SpecialName()> Public Sub DeleteMember(ByVal name As String)
        If Not _instance_dict.ContainsKey(name) Then
            Dim msg As String = String.Format(_error, name)
            Throw New System.MissingMemberException(msg)
        End If
        _instance_dict.Remove(name)
    End Sub

End Class
